reactjs-todo-dv/client/components/davinci-client/davinci-flow.js

```diff
@@ -0,0 +1,180 @@
+/*
+ * forgerock-sample-web-react
+ *
+ * davinci-flow.js
+ *
+ * Copyright (c) 2021 ForgeRock. All rights reserved.
+ * This software may be modified and distributed under the terms
+ * of the MIT license. See the LICENSE file for details.
+ */
+import React, { useContext, useEffect, useState, useRef } from 'react';
+import { useNavigate } from 'react-router-dom';
+import { TokenManager, UserManager } from '@forgerock/javascript-sdk';
+import davinci from '@forgerock/davinci-client';
+import TextInput from './text-input.js';
+import Password from './password.js';
+import SubmitButton from './submit-button.js';
+import Protect from './protect.js';
+import FlowButton from './flow-button.js';
+import ErrorMessage from './error-message.js';
+import { AppContext } from '../../global-state.js';
+
+/**
+ * @function DaVinciFlow - React view for a DaVinci flow
+ * @returns {Object} - React component object
+ */
+export default function DaVinciFlow({ config }) {
+  /**
+   * Collects the global state for detecting user auth for rendering
+   * appropriate navigational items.
+   * The destructing of the hook's array results in index 0 having the state value,
+   * and index 1 having the "setter" method to set new state values.
+   */
+  const [, methods] = useContext(AppContext);
+  const [components, setComponents] = useState([]);
+  const [pageHeader, setPageHeader] = useState('');
+
+  // Used for redirection after success
+  const navigate = useNavigate();
+
+  let client = useRef(null);
+
+  const renderComplete = async (successNode) => {
+    const code = successNode.client?.authorization?.code || '';
+    const state = successNode.client?.authorization?.state || '';
+    await TokenManager.getTokens({ query: { code, state } });
+    const user = await UserManager.getCurrentUser();
+    methods.setUser(user.preferred_username);
+    methods.setEmail(user.email);
+    methods.setAuthentication(true);
+    navigate('/');
+  };
+
+  const renderError = (errorNode) => {
+    const errorMessage = <ErrorMessage message={errorNode.error.message} key="error-message" />;
+    setComponents([errorMessage]);
+  };
+
+  // Represents the main render function for app
+  async function renderForm(nextNode) {
+    const arrFormFields = [];
+    // clear form contents
+    setComponents([]);
+    // Set h1 header
+    setPageHeader(nextNode.client?.name || '');
+
+    const collectors = client.current.collectors();
+
+    collectors.forEach((collector) => {
+      if (collector.type === 'TextCollector' && collector.name === 'protectsdk') {
+        arrFormFields.push(
```

Comment:
I feel like the combination of `Array.forEach` and the `Array.push` within it is bit too "mutation driven". If we did something like this, I'd encourage the use of `Array.map` that returns a new array, rather than mutating an existing array with `forEach`. I hope that makes sense. When I'm using React, I think the more Functional Programming approach with its principle of immutability is best.